### ISSUE TITLE
Support for complex numbers?

### ISSUE THREAD

**Issue by twesterhout** [VcDevel/Vc#197]
A question: are there any plans to add support for complex numbers?

For example, `Vc::simd<std::complex<float>, Vc::simd_abi::sse>` would still use `__m128` as underlying type, but `size()` would return 2 rather than 4.

**Comment by mattkretz** [comment]
I'd like to support simd<T> for user-defined T at some point when the C++ reflection capabilities are good enough. However, what you're really looking for is std::complex<Vc::simd<float>>. It almost works, except for the places where the implementation requires compares to return something boolean. I believe the required changes to std::complex, to make it work all the way, are not very large. Once we now what it takes, I'd like to propose simd support for std::complex inside WG21. If you're interested in helping out, let me know.

**Comment by twesterhout** [comment]
Not really, `std::complex<Vc::simd<T,...>>` would give me a SoA layout (or am I missing something?). Nowadays, however, most linear algebra libraries, for example, use AoS layout for complex numbers. So I really want `Vc::simd<std::complex<T>,...>` to e.g. pass in to functions like https://software.intel.com/sites/landingpage/IntrinsicsGuide/#!=undefined&text=clog&expand=661 without extra shuffling.

P.S. Definitely interested in helping out with the proposal.

**Comment by mattkretz** [comment]
True, but consider that std::log/exp/sin/... are defined on std::complex<T> and thus would automatically exist for std::complex<simd<T>>. Intel certainly did some heroics to support simd<complex> in their instruction set. But I believe that effort is misguided. I.e. one simd object should store only values that are the same scalar variable in the math formula, just filled from different input data. Thus real and imag need to be in different simd objects. So
```cpp
struct complex {
  simd<float> real, imag;
};
```
is, in my experience, the best you can get for speeding up your computation. Because that way you won't need any horizontal operations (which effectively reduce you're achievable parallelism by >=2). I understand your argument for compatibility to existing Intel APIs. But let's just force Intel to support complex<simd> by blessing it through the C++ standard. ;-)

The only thing complex<simd<T>> doesn't support without some extra help, is the transition between complex<T> and complex<simd<T>> objects. That's the main reason for me to invest into simd<UDT>. This would also create a struct of simd objects. For a proof of concept you can look at Vc::simdize<T> in Vc 1.3.
